### PR TITLE
Feat/scalar with window

### LINKED ISSUE #1540: [FEATURE] Scaler with rolling/expanding window to eliminate look ahead bias
**Problem description**
Currently, [Scaler](darts/dataprocessing/transformers/scaler.py) transforms input series "globally", meaning that all values of the the input vector are considered:

```python
from darts.dataprocessing.transformers import Scaler
from darts import TimeSeries

scaler = Scaler()

s = [1, 2, 3, 4, 5, 6, 7, 8, 9]
s = TimeSeries.from_series(s)

s_norm = scaler.fit_transform(s)
s_norm.pd_series()

0    0.000
1    0.125
2    0.250
3    0.375
4    0.500
5    0.625
6    0.750
7    0.875
8    1.000
```

This is not a problem if data is manually split into train and test sets and scaler is fitted with training set.

However, if we go on to use this vector as an input to [historical forecasts](https://unit8co.github.io/darts/quickstart/00-quickstart.html?highlight=historical_forecast#Backtesting:-simulate-historical-forecasting), we risk introducing [look ahead bias](https://stats.stackexchange.com/questions/442739/look-ahead-bias-induced-by-standardization-of-a-time-series) into analysis (at least when performing normalization). To eliminate this bias, rolling window approach is at least sometimes used https://arxiv.org/abs/1907.09452.

**Describe proposed solution**
One solution would be to add parameters to scaler like so:

```python
class Scaler(InvertibleDataTransformer, FittableDataTransformer):
    def __init__(
        self, scaler=None, type="Global", window=None, name="Scaler", n_jobs: int = 1, verbose: bool = False
    ):
   """
   Parameters
   ----------
   type
        The type to scale the data. Options:
        "Global" uses all data points 
        "Rolling" uses rolling window with window size specified by 'window' parameter
        "Expanding" uses expanding window with initial window size specified by 'window' parameter
    window
        Size of window if type is either "Rolling" or "Expanding"
    """
```

**Describe potential alternatives**
Another option is to integrate this functionality to historical_forecasts and backtest functions. This might be convenient because parameters above could be inferred by the desired backtesting setup.

**Additional context**
Thank you again for excellent software!

### PR DESCRIPTION
**Pull request by madtoinou**
Checklist before merging this PR:
- [x] Mentioned all issues that this PR fixes or addresses.
- [x] Summarized the updates of this PR under **Summary**.
- [x] Added an entry under **Unreleased** in the [Changelog](../CHANGELOG.md).

Fixes #1540.

### Summary

- Possibility to pass `Scaler/Pipeline` to `historical_forecasts()`, `backtest()` and `gridsearch()` to avoid data-leakage
  - When `enable_optimization=True`, the `Pipeline` must already be fitted (when applicable), all the series are transformed in one pass
  - `Pipeline` are systematically fitted and applied to the series between each forecast horizon (regardless of the `retrain` parameter value)

### Other Information

The input series to these methods must always be "un-processed" when providing `data_transformers` in order to avoid "double scaling" of the series.

This PR is based on #2021.

### CODE REVIEW

File: darts/models/forecasting/forecasting_model.py [hunk 1]
```diff
@@ -1053,6 +1080,19 @@ def retrain_func(
...
+                if len(data_transformers) > 0:
+                    # data transformers are retrained between iterations to avoid data-leakage
+                    train_series, past_covariates_, future_covariates_ = (
+                        _apply_data_transformers(
+                            series=train_series,
+                            past_covariates=past_covariates_,
+                            future_covariates=future_covariates_,
+                            data_transformers=data_transformers,
+                            max_future_cov_lag=model.extreme_lags[5],
+                            fit_transformers=True,
```
Comment (dennisbader) [inline darts/models/forecasting/forecasting_model.py:1092]: this should be defined according to `retrain`, right?

File: darts/models/forecasting/forecasting_model.py [hunk 2]
```diff
@@ -1193,6 +1241,9 @@ def backtest(
         reduction: Union[Callable[..., float], None] = np.mean,
         verbose: bool = False,
         show_warnings: bool = True,
+        data_transformers: Optional[
```
Comment (dennisbader) [inline darts/models/forecasting/forecasting_model.py:1244]: add description

File: darts/models/forecasting/forecasting_model.py [hunk 3]
```diff
@@ -1533,6 +1585,9 @@ def gridsearch(
         verbose=False,
         n_jobs: int = 1,
         n_random_samples: Optional[Union[int, float]] = None,
+        data_transformers: Optional[
```
Comment (dennisbader) [inline darts/models/forecasting/forecasting_model.py:1588]: add description

File: darts/models/forecasting/forecasting_model.py [hunk 4]
```diff
@@ -1773,27 +1838,49 @@ def _evaluate_combination(param_combination) -> float:
                     last_points_only=last_points_only,
                     verbose=verbose,
                     show_warnings=show_warnings,
+                    data_transformers=data_transformers,
                     fit_kwargs=fit_kwargs,
                     predict_kwargs=predict_kwargs,
                     sample_weight=sample_weight,
                 )
             else:  # split mode
+                if len(data_transformers) > 0:
```
Comment (dennisbader) [inline darts/models/forecasting/forecasting_model.py:1847]: should we also apply this to the `use_fitted_values=True` case?

File: darts/models/forecasting/forecasting_model.py [hunk 5]
```diff
@@ -1716,6 +1771,16 @@ def gridsearch(
                 logger,
             )
 
+        if data_transformers is None:
```
Comment (dennisbader) [inline darts/models/forecasting/forecasting_model.py:1774]: could be nice to have a utils function for that somewhere

File: darts/models/forecasting/forecasting_model.py [hunk 6]
```diff
@@ -1773,27 +1838,49 @@ def _evaluate_combination(param_combination) -> float:
...
-                    past_covariates=past_covariates,
-                    future_covariates=future_covariates,
+                    series=series_,
+                    past_covariates=past_covariates_,
+                    future_covariates=future_covariates_,
                     num_samples=1,
                     verbose=verbose,
                     **predict_kwargs,
                 )
+                if (
```
Comment (dennisbader) [inline darts/models/forecasting/forecasting_model.py:1879]: also for this, a utils function would be nice, just to have the reoccuring logic covered in one function

File: darts/utils/historical_forecasts/optimized_historical_forecasts_regression.py [hunk 7]
```diff
@@ -173,6 +191,9 @@ def _optimized_historical_forecasts_last_points_only(
                 hierarchy=series_.hierarchy,
             )
         )
+    # single data transformer accross all series and forecasts, using optimized inverse_transform
```
Comment (dennisbader) [inline darts/utils/historical_forecasts/optimized_historical_forecasts_regression.py:194]:
```suggestion
    # single data transformer across all series and forecasts, using optimized inverse_transform
```

File: darts/utils/historical_forecasts/optimized_historical_forecasts_regression.py [hunk 8]
```diff
@@ -33,20 +37,34 @@ def _optimized_historical_forecasts_last_points_only(
...
+        data_transformers = dict()
     forecasts_list = []
     iterator = _build_tqdm_iterator(series, verbose)
     for idx, series_ in enumerate(iterator):
         past_covariates_ = past_covariates[idx] if past_covariates is not None else None
         future_covariates_ = (
             future_covariates[idx] if future_covariates is not None else None
         )
+        # Pipeline/DataTransformer must already be fitted, transform everything in one go
+        if len(data_transformers) > 0:
```
Comment (dennisbader) [inline darts/utils/historical_forecasts/optimized_historical_forecasts_regression.py:58]: I think the transformation should be applied on all series together, and not individually.

Also, in my opinion, we can move the transform / inverse transform out of this function and into `ForecastingModel.historical_forecasts()`. Apply transform before calling `model_optimized_historical_forecasts()`, and inverse transform before returning.

That would simplify things quite a bit

File: darts/tests/models/forecasting/test_global_forecasting_models.py [hunk 9]
```diff
@@ -815,3 +821,122 @@ def test_residuals(self):
...
+        "params",
+        product(
+            [
+                (
+                    {"series": sine_3_ts + 3},
+                    {"target": Scaler()},
+                ),
+                (
+                    {"series": sine_3_ts + 3, "past_covariates": sine_1_ts + 3},
+                    {"past": Scaler(scaler=MinMaxScaler())},
```
Comment (dennisbader) [inline darts/tests/models/forecasting/test_global_forecasting_models.py:835]: That is the default already

```suggestion
                    {"past": Scaler()},
```

File: darts/tests/models/forecasting/test_global_forecasting_models.py [hunk 10]
```diff
@@ -815,3 +821,122 @@ def test_residuals(self):
...
+                        "past_covariates": sine_2_ts + 3,
+                        "future_covariates": sine_1_ts + 3,
+                    },
+                    {"target": Scaler(), "past": Scaler()},
+                ),
+            ],
+            [True, False],
+        ),
+    )
+    def test_historical_forecasts_with_scaler(self, params):
```
Comment (dennisbader) [inline darts/tests/models/forecasting/test_global_forecasting_models.py:853]: nice one 👍  :) 
Could you check for 2 historical forecast iterations? Also checking `last_points_only=False`, and `enable_optimization=True` would be important.

**Review by dennisbader — CHANGES_REQUESTED** [review]
Thanks a lot @madtoinou this looks really great 🚀  I like the approach with the Pipeline instead of Transformers :) 

Had some remarks and suggestions, we can discuss later.

File: darts/dataprocessing/pipeline.py [hunk 11]
```diff
@@ -217,6 +221,32 @@ def invertible(self) -> bool:
         """
         return self._invertible
 
+    def fittable(self) -> bool:
```
[the remark this answers is not in the document]
Comment (madtoinou) [reply darts/dataprocessing/pipeline.py:224]: I agree, I did not like the fact that API was different for the pipeline (compared to models for example)

File: darts/utils/historical_forecasts/utils.py [hunk 12]
```diff
@@ -211,6 +212,23 @@ def _historical_forecasts_general_checks(model, series, kwargs):
                 logger,
             )
 
+    # ensure DataTransformer are already be fitted when using the optimized historical forecasts
+    if isinstance(n.data_transformers, dict) and n.enable_optimization:
```
[the remark this answers is not in the document]
Comment (madtoinou) [reply darts/utils/historical_forecasts/utils.py:216]: I reused the naming convention of the lags, but probably better to go with the name of the series directly.

File: darts/models/forecasting/forecasting_model.py (hunk 4 above)
Comment (madtoinou) [reply darts/models/forecasting/forecasting_model.py:1847]: `fitted_values` is actually supported only by very models, I am not sure that it's relevant to add data transformer here... But going to do it, just in case

File: darts/utils/historical_forecasts/optimized_historical_forecasts_regression.py (hunk 8 above)
Comment (madtoinou) [reply darts/utils/historical_forecasts/optimized_historical_forecasts_regression.py:58]: I thought about it, decided to make it redundant in case some people want to call the optimized routine directly but I don't mind moving most of the logic to `historical_forecasts` and have only the inverse transform occur in the optimized methods.

Comment (dennisbader) [reply darts/utils/historical_forecasts/optimized_historical_forecasts_regression.py:58]: I would say we can leave the optimized routines naive to transformations. Also, applying it on all series/hist fc output is more efficient due to parallelization :)

**Comment by madtoinou** [comment]
<p>I added it to hide the warnings, but I don't mind removing it.</p>

---

 <a href='https://app.reviewnb.com/unit8co/darts/pull/2529/discussion/#comment-2470029056'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by madtoinou** [comment]
<p>Updated the notebook; two examples (one with, one without the auto-scaling) while minimizing the changes</p>

---

 <a href='https://app.reviewnb.com/unit8co/darts/pull/2529/discussion/#comment-2470029094'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

File: darts/tests/utils/historical_forecasts/test_historical_forecasts.py [hunk 13]
```diff
@@ -2630,6 +2648,342 @@ def test_probabilistic_optimized_hist_fc_regression(self, config):
...
+                    },
+                    {"series": Scaler(), "past_covariates": Scaler()},
+                ),
+            ],
+            [True, False],  # retrain
+            [True, False],  # last point only
+            models,
+        ),
+    )
+    def test_historical_forecasts_with_scaler(self, params):
```
Comment (dennisbader) [inline darts/tests/utils/historical_forecasts/test_historical_forecasts.py:2769]: nice test 👍

File: darts/tests/utils/historical_forecasts/test_historical_forecasts.py [hunk 14]
```diff
@@ -112,6 +123,8 @@
 
     NB_EPOCH = 1
 
+    models += [NLinearModel, RNNModel]
```
Comment (dennisbader) [inline darts/tests/utils/historical_forecasts/test_historical_forecasts.py:126]: one torch model should be enough

```suggestion
    models += [NLinearModel]
```

File: darts/tests/utils/historical_forecasts/test_historical_forecasts.py [hunk 15]
```diff
@@ -2630,6 +2648,342 @@ def test_probabilistic_optimized_hist_fc_regression(self, config):
...
+                columns=manual_hf_0.components,
+            )
+            self.helper_compare_hf(tmp_ts, hf_auto, last_points_only)
+        else:
+            self.helper_compare_hf(
+                hf_auto, [manual_hf_0, manual_hf_1], last_points_only
+            )
+
+    def test_historical_forecasts_with_scaler_errors(self):
+        """"""
```
Comment (dennisbader) [inline darts/tests/utils/historical_forecasts/test_historical_forecasts.py:2860]: missing docstring

File: darts/tests/utils/historical_forecasts/test_historical_forecasts.py [hunk 16]
```diff
@@ -2630,6 +2648,342 @@ def test_probabilistic_optimized_hist_fc_regression(self, config):
...
+                data_transformers={"series": Scaler()},
+                retrain=False,
+            )
+        assert str(err.value).startswith(
+            "All the fittable entries in `data_transformers` must already be fitted when `retrain=False`."
+        )
+
+        # retrain=False and data transformers fitted on multplie series with fit_global=False
+        with pytest.raises(ValueError) as err:
+            tmp_scaler = Scaler(global_fit=False)
```
Comment (dennisbader) [inline darts/tests/utils/historical_forecasts/test_historical_forecasts.py:2886]: we need to add support for this (see my previous comment in `darts/utils/historical_forecasts/utils.py`)

File: darts/tests/utils/historical_forecasts/test_historical_forecasts.py [hunk 17]
```diff
@@ -2630,6 +2648,342 @@ def test_probabilistic_optimized_hist_fc_regression(self, config):
...
+            )
+            assert len(opti_hf_auto) == len(hf_auto) == 2
+            # check that all the nested forecasts are identicals
+            for hf_list, opti_hf_list in zip(hf_auto, opti_hf_auto):
+                for fc_auto, fc_opti in zip(hf_list, opti_hf_list):
+                    assert fc_auto.time_index.equals(fc_opti.time_index)
+                    np.testing.assert_almost_equal(
+                        fc_auto.values(),
+                        fc_opti.values(),
+                    )
```
Comment (dennisbader) [inline darts/tests/utils/historical_forecasts/test_historical_forecasts.py:2985]: here also

File: darts/tests/utils/historical_forecasts/test_historical_forecasts.py [hunk 18]
```diff
@@ -2630,6 +2648,342 @@ def test_probabilistic_optimized_hist_fc_regression(self, config):
...
+        manual_hf_1_1 = self.helper_manual_scaling_prediction(
+            model, {"series": series[1]}, hf_scaler, retrain, -ocl, ocl
+        )
+
+        # check that all the nested forecasts are identicals
+        for auto_hf_list, manual_hf_list in zip(
+            hf_auto, [[manual_hf_0_0, manual_hf_0_1], [manual_hf_1_0, manual_hf_1_1]]
+        ):
+            for fc_auto, fc_manual in zip(auto_hf_list, manual_hf_list):
+                assert fc_auto.time_index.equals(fc_manual.time_index)
```
Comment (dennisbader) [inline darts/tests/utils/historical_forecasts/test_historical_forecasts.py:2961]: you could use your hfc compare helper from above

File: CHANGELOG.md [hunk 19]
```diff
@@ -12,6 +12,7 @@ but cannot always guarantee backwards compatibility. Changes that may **break co
 **Improved**
 
 - Improvements to `ForecastingModel`: Improved `start` handling for historical forecasts, backtest, residuals, and gridsearch. If `start` is not within the trainable / forecastable points, uses the closest valid start point that is a round multiple of `stride` ahead of start. Raises a ValueError, if no valid start point exists. This guarantees that all historical forecasts are `n * stride` points away from start, and will simplify many downstream tasks. [#2560](https://github.com/unit8co/darts/issues/2560) by [Dennis Bader](https://github.com/dennisbader).
+- Added `data_transformers` argument to `historical_forecasts`, `backtest` and `gridsearch` that allows scaling of the series without data-leakage. [#2529](https://github.com/unit8co/darts/pull/2529) by [Antoine Madrona](https://github.com/madtoinou) and [Jan Fidor](https://github.com/JanFidor)
```
Comment (dennisbader) [inline CHANGELOG.md:15]: `residuals()` support is missing

**Review by dennisbader — COMMENTED** [review]
This looks really good @madtoinou 🚀 

I added a couple of suggestions, and two things which we should add:
- support data transformers in residuals
- support pre-trained data transformers regardless of `global_fit`

File: darts/utils/historical_forecasts/utils.py [hunk 20]
```diff
@@ -193,6 +198,97 @@ def _historical_forecasts_general_checks(model, series, kwargs):
...
+                            f"When multiple series are provided, their number should match the number of "
+                            f"`TimeSeries` used to fit the data transformers `n={max(fitted_params_pipelines)}` "
+                            f"(only relevant for fittable transformers that use `global_fit=False`)."
+                        ),
+                        logger,
+                    )
+            else:
+                # at least one pipeline was fitted on several series with `global_fit=False` but only
+                # one series was passed
+                if max(fitted_params_pipelines) > 1:
```
Comment (dennisbader) [inline darts/utils/historical_forecasts/utils.py:286]: ignore this when `ignore_warnings=True`

**Review by dennisbader — CHANGES_REQUESTED** [review]
Beautiful 😍 I pushed some minor changes:
- adapt the docs a bit here and there
- renamed `idx_series` to `series_idx`
- converted `transform()` of fitted_values (in gridsearch) to `inverse_transform()` (should be inverse transform I think ;) )

There is one last change that we throw a warning instead of an error when `global_fit=True` and `len(series)>1`

**Review by dennisbader — APPROVED** [review]
Great work @madtoinou and @JanFidor 🚀 Thanks for all the work!
